[WIP] Feat/state lang

naterush opened the pull request:
ADDS:
- Moves core functionality to StateLanguage.
- Testing languages for specific blockchains that inherit from StateLanguage.
- Refactors lots of tests to deal w/ these breaking changes

naterush commented:
I'm not sure testing_langauges is the best place for these to be. Happy to put them in the protocols, if you think that makes sense @djrtwo (I think it might). Let me know.

naterush commented:
@djrtwo same goes for above. Whatever you think is best in regards to organization.

naterush commented:
@djrtwo fixed the tests, but not sure the way I'm testing currently makes the most sense. In many of the tests, loop through all the protocols and test each of them. This seemed like the most reasonable strategy, rather than passing each protocol as a parameter to the tests. 

There might be some better pytest method of doing this, which I'll look into, but let me know if you have any thoughts.

naterush commented:
Also, I'm not totally comfortable with how the TestingLang and StateLang are split up. Maybe it makes sense to have a "translate" function, which takes a command like `SJ` and totally removes it (by turning it into a string of `S`). I think this might be a more pure way of doing things.

naterush commented:
The above requires a reverse mapping of messages => message_names. I could also implement a two-way dictionary (something like https://stackoverflow.com/questions/1456373/two-way-reverse-map). Alternatively, could use something like https://pypi.python.org/pypi/bidict. For now, going to just keep a reverse mapping.

naterush commented:
The translation hell (as well as the complicated inheritance) makes me think the best option here is increasing the complexity of the `StateLang` and just adding `RR` and `SJ` directly to the `StateLang`. 

While this is a less pure (maybe), it makes things a lot less complicated in both respects above. Furthermore, this `StatLang` can contain the unimplemented function stubs for assertions that are currently found in the `TestLang`, and we can just have a single level of inheritance for implementing those (which I think is unavoidable).

naterush commented:
@djrtwo your comments about the complicated inheritance, as well as the translation hell I was dealing with, led me removing the `TestingLang` and adding `RR` and `SJ` directly to the `StateLang`. 

While this is a less pure (maybe), it makes things a lot less complicated in both respects above. Furthermore, this `StatLang` can contains the unimplemented function stubs for assertions that were found in the `TestLang`, and we now we just have a single level of inheritance for implementing those (which I think is unavoidable).

djrtwo commented on tests/casper/langs/test_binary_test_lang.py:
```diff
@@ -0,0 +1,134 @@
+
+    for validator in binary_lang.validator_set:
+        assert validator.view.last_finalized_estimate is None
+
```
two newlines here

djrtwo commented on tests/casper/langs/test_binary_test_lang.py:
```diff
@@ -0,0 +1,134 @@
+    binary_lang.parse(check_unsafe)
+
+
+def test_check_unsafe_fails_fails_on_invalid_assertions(test_weight):
```
two `fails` in method name. Probably just need one

djrtwo commented on tests/casper/langs/test_binary_test_lang.py:
```diff
@@ -0,0 +1,134 @@
+
+    current_estimate = binary_lang.network.global_view.estimate()
+
+    with pytest.raises(AssertionError):
```
If we have more deterministic initial_messages, this could be updated with the known estimate.

djrtwo commented on tests/casper/langs/test_blockchain_test_lang.py:
```diff
@@ -0,0 +1,99 @@
+
+
+def test_check_estimate_passes_on_valid_assertions(test_weight):
+    blockchain_lang = BlockchainTestLang(test_weight, False)
```
Same as `binary_lang`, define and use a `blockchain_lang` fixture throughout.

djrtwo commented on tests/casper/langs/test_blockchain_test_lang.py:
```diff
@@ -0,0 +1,99 @@
+        ('RR0-A RR0-B RR0-C RR0-D RR0-E RR0-F CS1-F'),
+    ]
+)
+def test_check_safe_fails_fails_on_invalid_assertions(test_weight, test_string):
```
methodname has two "fails" should probably have one

djrtwo commented on tests/casper/langs/test_blockchain_test_lang.py:
```diff
@@ -0,0 +1,99 @@
+        ('M0-A RR0-B RR0-C RR0-D CE1-A CS1-A CU0-A'),
+    ]
+)
+def test_check_unsafe_fails_fails_on_invalid_assertions(test_weight, test_string):
```
two "fails" in methodname. should have one

djrtwo commented on tests/casper/langs/test_integer_test_lang.py:
```diff
@@ -0,0 +1,136 @@
+
+
+def test_only_accepts_integer_estimates(test_weight):
+    integer_lang = IntegerTestLang(test_weight, False)
```
Define as a fixture

djrtwo commented on tests/casper/langs/test_integer_test_lang.py:
```diff
@@ -0,0 +1,136 @@
+    integer_lang = IntegerTestLang(test_weight, False)
+    integer_lang.parse('M0-A')
+
+    with pytest.raises(ValueError):
```
Use `parametrize`

djrtwo commented on tests/casper/langs/test_integer_test_lang.py:
```diff
@@ -0,0 +1,136 @@
+        ('RR0-A RR0-B CS1-0 CS1-1'),
+    ]
+)
+def test_check_safe_fails_fails_on_invalid_assertions(test_weight, test_string):
```
fix name. should only have one "fails"

djrtwo commented on tests/casper/langs/test_integer_test_lang.py:
```diff
@@ -0,0 +1,136 @@
+
+    for validator in integer_lang.validator_set:
+        assert validator.view.last_finalized_estimate is None
+
```
two spaces

djrtwo commented on tests/casper/langs/test_state_lang.py:
```diff
@@ -0,0 +1,292 @@
+    state_lang = StateLanguage(test_weight, protocol, False)
+
+    assert isinstance(state_lang.network, Network)
+
```
only two new lines

djrtwo commented on tests/casper/langs/test_state_lang.py:
```diff
@@ -0,0 +1,292 @@
+INITAL_MESSAGE_PROTOCOLS = [BinaryProtocol, OrderProtocol, IntegerProtocol]
+
+
+@pytest.mark.parametrize(
```
define `protocol_params` at the top of the file:

`protocol_params = [(protocol) for protocol in PROTOCOLS]`

 and then just use this line each time you want to pass in just protocols to a test:

`@pytest.mark.parametrize('protocol', protocol_params)`

djrtwo commented on tests/casper/langs/test_state_lang.py:
```diff
@@ -0,0 +1,292 @@
+    ]
+)
+def test_parse_only_valid_tokens_split_strings(test_strings, test_weight, error):
+    for protocol in PROTOCOLS:
```
Can possibly build the parametrize data with a loop and then pass in "protocol, test_strings, error". Will help out if there happens to be a test broken against one protocol but not the others.

djrtwo replied:
Same for tests below. If it's too messy, don't worry about it.

djrtwo commented on tests/casper/langs/test_state_lang.py:
```diff
@@ -0,0 +1,292 @@
+        assert len(state_lang.network.global_view.justified_messages) == num_blocks + num_inital_blocks
+
+
+# NOTE: None means the block is not named by the testing language
```
Is this comment relevant? Not seeing any reference to `None` in the tests

djrtwo commented on tests/casper/protocols/blockchain/test_block.py:
```diff
@@ -17,18 +17,19 @@ def test_equality_of_copies_off_genesis(validator, empty_just):
 def test_equality_of_copies_of_non_genesis(report):
-    test_string = "B0-A S1-A B1-B S0-B B0-C S1-C B1-D S0-D H0-D"
-    test_lang = TestLangCBC({0: 10, 1: 11}, BlockchainProtocol, report)
+    pass
```
Don't think you need `pass` if you have the `mark.skip` above

naterush commented on tests/casper/langs/test_binary_test_lang.py:
```diff
@@ -0,0 +1,134 @@
+    binary_lang = BinaryTestLang(test_weight, False)
+    binary_lang.parse('M0-A S1-A S2-A S3-A S4-A')
+
+    current_estimates = dict()
```
Yeah, it would be good to specify this. For now, I'm going to open an issue and we can discuss how to best address this.

naterush commented:
@djrtwo addressed most of the comments above, with the exception of parameterizing protocols in the way you describe (seemed to make things more complicated, but maybe I'm missing something).

Would appreciate a final review, and then good to merge :)

djrtwo commented:
excellent. merging
